File: presentation/src/main/java/ru/andrey/cleandictionary/presentation/view/list/WordAdapter.java
package ru.andrey.cleandictionary.presentation.view.list;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.v7.recyclerview.extensions.ListAdapter;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.List;

import ru.andrey.cleandictionary.R;
import ru.andrey.cleandictionary.presentation.dto.TranslationDto;

public class WordAdapter extends ListAdapter<TranslationDto, WordAdapter.Holder> {

    private ItemClicked mItemClicked;

    private Drawable[] stars = new Drawable[2];

    WordAdapter(ItemClicked itemClicked, Context context) {
        super(DIFF_CALLBACK);
        mItemClicked = itemClicked;
        stars[0] = context.getDrawable(R.drawable.ic_no_star_24dp);
        stars[1] = context.getDrawable(R.drawable.ic_star_24dp);
    }

    @NonNull
    @Override
    public Holder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.dictionary_list_item, parent, false);
        return new Holder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull Holder holder, int position) {
        final TranslationDto item = getItem(position);
        holder.header.setText(item.getWord().getWord());
        holder.translation.setText(item.getTranslation());
        holder.langFrom.setText(item.getWord().getFrom());
        holder.langTo.setText(item.getWord().getTo());
        holder.star.setImageDrawable(getStarImage(item.getFavorite()));

        holder.star.setOnClickListener(v -> mItemClicked.onStarClicked(item));
    }

    @Override
    public void onBindViewHolder(@NonNull Holder holder, int position, @NonNull List<Object> payloads) {
        if (payloads.isEmpty()) {
            onBindViewHolder(holder, position);
        }
        for (Object payload : payloads) {
            switch ((ItemPayload) payload) {
                case FAVORITE:
                    holder.star.setImageDrawable(getStarImage(getItem(position).getFavorite()));
            }
        }
    }

    private Drawable getStarImage(boolean enabled) {
        return enabled ? stars[1] : stars[0];
    }

    static class Holder extends RecyclerView.ViewHolder {

        TextView header;
        TextView translation;
        ImageView star;
        TextView langFrom;
        TextView langTo;
        CardView card;

        Holder(View itemView) {
            super(itemView);
            header = itemView.findViewById(R.id.word_header);
            translation = itemView.findViewById(R.id.word_translation);
            star = itemView.findViewById(R.id.star_image);
            langFrom = itemView.findViewById(R.id.lang_from);
            langTo = itemView.findViewById(R.id.lang_to);
            card = itemView.findViewById(R.id.word_card);
        }
    }

    private static final DiffUtil.ItemCallback<TranslationDto> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<TranslationDto>() {
                @Override
                public boolean areItemsTheSame(TranslationDto oldDto, TranslationDto newDto) {
                    return oldDto.getId() == newDto.getId();
                }

                @Override
                public boolean areContentsTheSame(TranslationDto oldDto, TranslationDto newDto) {
                    return oldDto.equals(newDto);
                }

                @Override
                public Object getChangePayload(TranslationDto oldItem, TranslationDto newItem) {
                    if (oldItem.getFavorite() != newItem.getFavorite()) {
                        return ItemPayload.FAVORITE;
                    }
                    return super.getChangePayload(oldItem, newItem);
                }
            };

    enum ItemPayload {
        FAVORITE
    }

    interface ItemClicked {
        void onStarClicked(TranslationDto translation);
    }
}

File: presentation/src/main/java/ru/andrey/cleandictionary/presentation/view/list/WordAdapter.kt
package ru.andrey.cleandictionary.presentation.view.list

import android.content.Context
import android.graphics.drawable.Drawable
import android.support.v7.recyclerview.extensions.ListAdapter
import android.support.v7.util.DiffUtil
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import ru.andrey.cleandictionary.R
import ru.andrey.cleandictionary.presentation.dto.TranslationDto
import ru.andrey.cleandictionary.presentation.view.list.WordAdapter.ItemPayload.FAVORITE

class WordAdapter(private val itemClicked: (Int) -> Unit,
                  context: Context) : ListAdapter<TranslationDto, Holder>(DiffCallback()) {

    private val stars = arrayOf(
            context.getDrawable(R.drawable.ic_no_star_24dp),
            context.getDrawable(R.drawable.ic_star_24dp))

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Holder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.dictionary_list_item, parent, false)
        return Holder(view)
    }

    override fun onBindViewHolder(holder: Holder, position: Int) {
        val item = getItem(position)

        holder.id = item.id
        holder.header.text = item.word.word
        holder.translation.text = item.translation
        holder.langFrom.text = item.word.from
        holder.langTo.text = item.word.to
        holder.star.setImageDrawable(getStarImage(item.favorite))

        holder.star.setOnClickListener { itemClicked(item.id) }
    }

    override fun onBindViewHolder(holder: Holder, position: Int, payloads: List<Any>) {
        if (payloads.isEmpty()) {
            onBindViewHolder(holder, position)
        }
        for (payload in payloads) {
            when (payload as ItemPayload) {
                FAVORITE -> holder.star.setImageDrawable(getStarImage(getItem(position).favorite))
            }
        }
    }

    override fun getItemId(position: Int): Long {
        return getItem(position).id.toLong()
    }

    private fun getStarImage(enabled: Boolean): Drawable {
        return if (enabled) stars[1] else stars[0]
    }

    internal enum class ItemPayload {
        FAVORITE
    }
}

class Holder(itemView: View) : RecyclerView.ViewHolder(itemView) {

    var id: Int = 0
    var header: TextView = itemView.findViewById(R.id.word_header)
    var translation: TextView = itemView.findViewById(R.id.word_translation)
    var star: ImageView = itemView.findViewById(R.id.star_image)
    var langFrom: TextView = itemView.findViewById(R.id.lang_from)
    var langTo: TextView = itemView.findViewById(R.id.lang_to)
}

internal class DiffCallback : DiffUtil.ItemCallback<TranslationDto>() {

    override fun areItemsTheSame(oldDto: TranslationDto, newDto: TranslationDto): Boolean {
        return oldDto.id == newDto.id
    }

    override fun areContentsTheSame(oldDto: TranslationDto, newDto: TranslationDto): Boolean {
        return oldDto == newDto
    }

    override fun getChangePayload(oldItem: TranslationDto, newItem: TranslationDto): Any {
        return if (oldItem.favorite != newItem.favorite) {
            FAVORITE
        } else super.getChangePayload(oldItem, newItem)
    }
}
